### DISCUSSION TITLE
STM32L476RE based Disco board not running Micropython

### DISCUSSION THREAD

**Discussion by kowshik1234** [micropython/micropython#13336] [STM32 / Pyboard] [answered]
Hi,

I have compiled the micropython from the source by making changes to the UART port on which REPL should launch and it generated a .hex file. 

<img width="769" alt="image" src="https://github.com/micropython/micropython/assets/26191169/294dc4b8-2c84-42a0-9adf-0751b00c7ccc">

I want to flash this to the MCU but I want to do it with STLink/V2 debugger. I tried flashing it but I don't see the REPL running on this UART port. What is the recommended way to flash the micropython hex files onto STM32L476RE MCU?

FYI, I have this Eval board - [ST25R3911B-DISCO - Discovery kit for the ST25R3911B high performance HF reader / NFC initiator - STMicroelectronics](https://www.st.com/en/evaluation-tools/st25r3911b-disco.html)

I am trying to do a porting of the NFC libraries of this board to micropython and as a part of my contribution task I tried doing the above. Please help me how to tweak and run the micropython.

Thanks,
G Kowshik

**Comment by davefes** [comment]
In my attempts to get REPL working on a STM32WLE5 I was told that REPL uses the LPUART1_TX and LPUART1_RX pins.  I then use a ST-Link V2, STM32 Cube Programmer and the SWCLK and SWDIO pins to program it.

**Reply by kowshik1234** [reply]
@davefes I tried using ST-LINK utility by STM itself and the flashing was success. Would there be a difference in the STM33 Cube programmer GUI vs ST-LINK utility GUI? I believe both should ultimately flash the code anyway.

And also, as I stated above I've changed the REPL UART instance so I expect it should use LPUART1 to boot repl.

**Reply by davefes** [reply]
Sorry, haven't used the ST-LINK utility GUI, but if it completes the task it possibly doing the same job as STM32 Cube.
I changed it to the two above the ones that you have high-lighted.
```py
// UART buses
#define MICROPY_HW_UART1_TX                     (pin_B6)    // Arduino D1, pin 7 on CN9
#define MICROPY_HW_UART1_RX                     (pin_B7)    // Arduino D0, pin 8 on CN9
#define MICROPY_HW_LPUART1_TX                   (pin_A2)    // Used for REPL STM32WLE5
#define MICROPY_HW_LPUART1_RX                   (pin_A3)    // Used for REPL STM32WLE5
#define MICROPY_HW_UART_REPL                    PYB_LPUART_1
#define MICROPY_HW_UART_REPL_BAUD               115200
```
Using a FTDI programmer on the changed UART pins.

**Reply by kowshik1234** [reply]
Yes I did exactly the same, but seems the REPL is not coming up. After flashing the hex if I reset the MCU, will it cause the REPL to boot again?

**Reply by davefes** [reply]
Have a look at discussion 13110 about the 32MHz oscillator ... it might be relevant.

**Reply by kowshik1234** [reply]
Thanks for the thread I went through it. I guess the suggestions given there won't help me much because the boards are very different, also the oscillators which you've told also doesn't apply to my case because my board solely works on the external 32Khz one so i don't need to bypass any oscillator configurations, however appreciate sharing the thread. I need some deeper debug on this, I'm hoping it'd be much simpler that I'm missing.

**Reply by kowshik1234** [reply]
An update - I have changed the LED pins as per my board and after I flash the hex file I can see the LED blinking. However I don't see the REPL booting up. Two questions

1. Will the LED blink and REPL parallely boot up? Or the LED blink happens only when the REPL launch fails?
2. How can I debug the C code with this hex? Can I use STMCube IDE, load the hex file and load the .elf for the symbols and debug it? What is the recommended way?

Thanks

**Reply by robert-hh** [reply]
If the LED. blink is done by Python code which runs at startup, then there will be no REPL prompt until that code terminates.

**Reply by kowshik1234** [reply]
@robert-hh I haven't written any python code yet. I'm using the STM32L476RG example port from the repo directly and changed the UART, LED pin instances only. How can I make the REPL boot instead of making the LED blink?

**Reply by robert-hh** [reply]
If you did not code for blinking yourself, the LED blinking indicates an error during boot.  Do you get any output at the serial port when it happens?

**Reply by kowshik1234** [reply]
@robert-hh Great, I wanna know if there's a particular pattern to indicate this error? Also what's the best way to debug the boot up process? Can I use STMCubeIDE and load the symbols from the elf? 

Thanks

**Reply by robert-hh** [reply]
Shuold be possible. I never debugged the STM32 port code, but for the other device the elf file contained all that was needed. It may be helpful to compile the firmware with DEBUG=1 at the make command line. Then optimization is disabled and it's easier to follow the source in the debugger.
The function that blinks the LED is boardctrl_fatal_error() in boardctrl.c. In the code you will see calls to MICROPY_BOARD_FATAL_ERROR(). If you set a breakpoint in boardctrl_fatal_error(), you can see in the stack trace from where it was called.

**Reply by kowshik1234** [reply]
@robert-hh Thanks for the insight, I have went ahead and started debugging. Here's what I found

1. Added a loop at the stm32_main function so that the CPU will be looping there forever until I load the symbols in the Cube IDE.
```
volatile uint8_t status = 1;

void loop_forever()
{
    while(status)
    {

    }
}

void stm32_main(uint32_t reset_mode) {
    // Low-level MCU initialisation.
    stm32_system_init();

    loop_forever();
```
2. Then I loaded the elf file in the STMCube, initiated the debug session. As expected the CPU stalled at the loop, I have loaded symbols and changed the status value from 1 to 0 in the expressions window to continue with the execution.

3. I followed the code and observed that it's checking for the user to press the User switch on the board, I used a jumper to create a button press manually. 

The below code present under the function `update_reset_mode` under the file `ports/stm32/boardctrl.c` is doing the switch press check.

```
        // For boards with only a single LED, we'll flash that LED the
        // appropriate number of times, with a pause between each one
        for (uint i = 0; i < 10; i++) {
            led_state(1, 0);
            for (uint j = 0; j < reset_mode; j++) {
                if (!switch_get()) {
                    break;
                }
                led_state(1, 1);
                HAL_Delay(100);
                led_state(1, 0);
                HAL_Delay(200);
            }
            HAL_Delay(400);
            if (!switch_get()) {
                break;
            }
            if (++reset_mode > BOARDCTRL_RESET_MODE_FACTORY_FILESYSTEM) {
                reset_mode = BOARDCTRL_RESET_MODE_NORMAL;
            }
        }
        // Flash the selected reset mode
        for (uint i = 0; i < 2; i++) {
            for (uint j = 0; j < reset_mode; j++) {
                led_state(1, 1);
                HAL_Delay(100);
                led_state(1, 0);
                HAL_Delay(200);
            }
            HAL_Delay(400);
        }
        #else
        #error Need a reset mode update method
        #endif
    }
```

4. Once I did it, I was able to proceed to the next step where the MCU started to read the flash contents, however in this process I got a `Hard Fault` and here's the stack & place where the hardfault occurred. (It happened during a memcpy)

<img width="742" alt="image" src="https://github.com/micropython/micropython/assets/26191169/e8975788-7709-4aa1-bb8b-af4135ab15ea">

As I debugged little deep, I understood that it's trying to copy from flash to internal RAM directly, 
<img width="557" alt="image" src="https://github.com/micropython/micropython/assets/26191169/1070e0eb-6136-4a2c-84ea-704d0d0e0e34">

the only reason I can think of this hardfault getting raised is if the flash is not being accessed correctly. If I remember, the flash must be in memmap(memory map) mode to read the data. Is this some issue with that? Please help me debug further.


Thanks,
G Kowshik

**Reply by robert-hh** [reply]
The code tries to read data from the file system at offset 0x80000 or 512k of the flash area. The STM32L476RE has 512k flash. So you try to read beyond the flash area of the MCU. It there other flash on the board, and at which address does it start?

**Reply by robert-hh** [reply]
You may have to adjust the loader file. The file boards/stm32l476xe.ld shows the memory areas as:
```
/* Specify the memory areas */
MEMORY
{
    FLASH (rx)      : ORIGIN = 0x08000000, LENGTH = 384K /* sectors 0-191 */
    FLASH_FS (r)    : ORIGIN = 0x08060000, LENGTH = 128K /* sectors 192-255 */
    RAM (xrw)       : ORIGIN = 0x20000000, LENGTH = 96K
    SRAM2 (xrw)     : ORIGIN = 0x10000000, LENGTH = 32K /* not contiguous with RAM */
}
```
Which loader script do you use?

**Reply by kowshik1234** [reply]
@robert-hh 

> Which loader script do you use?

I have enabled verbose build and from there I see the script is using the `stm32l476xg.ld` and it has the flash sections named as below

```
/* Specify the memory areas */
MEMORY
{
    FLASH (rx)      : ORIGIN = 0x08000000, LENGTH = 512K /* sectors 0-255 */
    FLASH_FS (r)    : ORIGIN = 0x08080000, LENGTH = 512K /* sectors 256-511 */
    RAM (xrw)       : ORIGIN = 0x20000000, LENGTH = 96K
    SRAM2 (xrw)     : ORIGIN = 0x10000000, LENGTH = 32K /* not contiguous with RAM */
}

_micropy_hw_internal_flash_storage_start = ORIGIN(FLASH_FS);
_micropy_hw_internal_flash_storage_end   = ORIGIN(FLASH_FS) + LENGTH(FLASH_FS);
```

**Reply by robert-hh** [reply]
It should use the stm32l476xe.ld script, matching the STM32L476RE device.

**Reply by kowshik1234** [reply]
@robert-hh I have taken a reference example from the STM's SDK and this is what I found from their linker script, seems the current one that is being picked by Micropython makesystem is correct only. What do you think please?

<img width="479" alt="image" src="https://github.com/micropython/micropython/assets/26191169/d91c2e19-e88a-4a17-b0bf-5268c69435f4">

Regardless, I will give a try by plugging in the `stm32l476xe.ld` values to the existing script and see how it goes. I remember trying this out and the execution gave a hard fault in the `mp_init( ) //Initialize micropython ` function, however let me try again.

Also on a lighter note, I'm newly getting this warning when build, Is it complaining about the linker file changes that has happened and the newly build binary has changes? I've no idea about this.

```
GEN build-NUCLEO_L476RG/frozen_content.c
CC build-NUCLEO_L476RG/frozen_content.c
LINK build-NUCLEO_L476RG/firmware.elf
   text    data     bss     dec     hex filename
 405828      96   27588  433512   69d68 build-NUCLEO_L476RG/firmware.elf
GEN build-NUCLEO_L476RG/firmware.bin
GEN build-NUCLEO_L476RG/firmware.hex
GEN build-NUCLEO_L476RG/firmware.dfu
make: warning:  Clock skew detected.  Your build may be incomplete.
```
Thanks

**Reply by robert-hh** [reply, marked as answer]
I do not know what the warning means. Sometimes it is the best to clean the build directory and make a complete new build.

As of the loader script defines: They tell, how much of the flash is used for code, and where the file system starts. The hard fault you have seen in debugging happened in accessing the file system in the FLASH_FS area, which according to your loader settings was outside of the existing flash. The stm32l476xe.ld settings move it into the existing flash range of 512k.

**Reply by kowshik1234** [reply]
@robert-hh Thanks alot for the help. I am able to boot up the REPL after changing the linker script to `xe.ld`, really appreciate your help. 

Thanks again, Cheers !!
